src/main/java/com/google/devtools/build/lib/query2/cquery/TransitionResolver.java

```diff
@@ -0,0 +1,168 @@
+import com.google.devtools.build.lib.analysis.DependencyResolver;
+import com.google.devtools.build.lib.analysis.InconsistentAspectOrderException;
+import com.google.devtools.build.lib.analysis.TargetAndConfiguration;
+import com.google.devtools.build.lib.analysis.ToolchainCollection;
+import com.google.devtools.build.lib.analysis.ToolchainContext;
+import com.google.devtools.build.lib.analysis.config.BuildConfigurationValue;
+import com.google.devtools.build.lib.analysis.config.BuildOptions;
+import com.google.devtools.build.lib.analysis.config.ConfigMatchingProvider;
+import com.google.devtools.build.lib.analysis.config.transitions.NoTransition;
+import com.google.devtools.build.lib.analysis.config.transitions.NullTransition;
+import com.google.devtools.build.lib.analysis.config.transitions.TransitionFactory;
+import com.google.devtools.build.lib.analysis.config.transitions.TransitionUtil;
+import com.google.devtools.build.lib.analysis.configuredtargets.RuleConfiguredTarget;
+import com.google.devtools.build.lib.cmdline.Label;
+import com.google.devtools.build.lib.concurrent.ThreadSafety.Immutable;
+import com.google.devtools.build.lib.events.ExtendedEventHandler;
+import com.google.devtools.build.lib.packages.RuleTransitionData;
+import com.google.devtools.build.lib.packages.Target;
+import com.google.devtools.build.lib.util.OrderedSetMultimap;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * TransitionResolver resolves the dependencies of a ConfiguredTarget, reporting which
+ * configurations its dependencies are actually needed in according to the transitions applied to
+ * them.
+ */
+public class TransitionResolver {
+
+  /**
+   * ResolvedTransition represents a single edge in the dependency graph, between some target
+   * and a target it depends on, reachable via a single attribute.
+   */
+  @AutoValue
+  @Immutable
+  public abstract static class ResolvedTransition {
+
+    static ResolvedTransition create(Label label, Collection<BuildOptions> configurationChecksum, String attributeName, String transitionName) {
+      return new AutoValue_TransitionResolver_ResolvedTransition(label, configurationChecksum, attributeName, transitionName);
+    }
+
+    /**
+     * The label of the target being depended on.
+     */
+    abstract Label label();
+
+    /**
+     * The configuration(s) this edge results in. This is a collection because a split transition
+     * may lead to a single attribute requesting a dependency in multiple configurations.
+     *
+     * If a target is depended on via two attributes, separate ResolvedTransitions should be used,
+     * rather than combining the two into a single ResolvedTransition with multiple options.
+     *
+     * If no transition was applied to an attribute, this collection will be empty.
+     */
+    abstract Collection<BuildOptions> options();
```

Comment:
Is it possible to use `ImmutableCollection` here instead because AutoValue instances should be deeply immutable?